File: leetcode/src/main/ruby/remove_nth_node.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
    dummy = ListNode.new(0)
    dummy.next = head
    l1, l2 = dummy, dummy
    (n + 1).times {l2 = l2.next}
    while l2
        l1, l2 = l1.next, l2.next
    end
    l1.next = l1.next.next
    return dummy.next
end

File: leetcode/src/main/golang/remove_nth_node.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    dummy := &ListNode{0, nil}
    dummy.Next = head
    l1, l2 := dummy, dummy
    for i := 0; i <= n; i++ {
        l2 = l2.Next
    }
    for ; l2 != nil; {
        l1, l2 = l1.Next, l2.Next
    }
    l1.Next = l1.Next.Next;
    return dummy.Next;
}